File: Metronome.java
// Metronome.java
import java.util.Scanner;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class Metronome {
    private int bpm;
    private int beatsPerMeasure;
    private int beatUnit;
    private int beatCount;
    private boolean running;
    private ScheduledExecutorService scheduler;

    public Metronome(int bpm, int beats, int unit) {
        this.bpm = bpm;
        this.beatsPerMeasure = beats;
        this.beatUnit = unit;
        this.beatCount = 0;
        this.running = false;
    }

    private void beep() {
        // Java has no simple beep; we'll print a bell symbol
        System.out.print(" 🔔");
    }

    private void visual(int beatNum, boolean isDownbeat) {
        char downbeatChar = '█';
        char otherChar = '░';
        char barChar = isDownbeat ? downbeatChar : otherChar;
        StringBuilder bar = new StringBuilder();
        for (int i = 0; i < 10; i++) bar.append(barChar);
        System.out.printf("\rBeat %d/%d: %s", beatNum, beatsPerMeasure, bar.toString());
    }

    private void tick() {
        if (!running) return;
        beatCount++;
        int beatNum = ((beatCount - 1) % beatsPerMeasure) + 1;
        boolean isDownbeat = beatNum == 1;
        visual(beatNum, isDownbeat);
        beep();
    }

    public void start() {
        running = true;
        beatCount = 0;
        long intervalMs = (long)(60000.0 / bpm);
        System.out.printf("\nStarting metronome at %d BPM (%d/%d). Press Ctrl+C to stop.\n", bpm, beatsPerMeasure, beatUnit);
        scheduler = Executors.newSingleThreadScheduledExecutor();
        scheduler.scheduleAtFixedRate(this::tick, 0, intervalMs, TimeUnit.MILLISECONDS);
        // Wait for interrupt
        try {
            Thread.sleep(Long.MAX_VALUE);
        } catch (InterruptedException e) {
            stop();
        }
    }

    public void stop() {
        if (!running) return;
        running = false;
        if (scheduler != null) {
            scheduler.shutdownNow();
            try { scheduler.awaitTermination(1, TimeUnit.SECONDS); } catch (InterruptedException e) {}
        }
        System.out.println("\nStopped.");
        System.exit(0);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("=== Metronome ===");
        System.out.print("Enter BPM (20-300) [120]: ");
        String bpmInput = scanner.nextLine().trim();
        int bpm = 120;
        if (!bpmInput.isEmpty()) {
            try {
                int v = Integer.parseInt(bpmInput);
                if (v >= 20 && v <= 300) bpm = v;
            } catch (NumberFormatException e) {}
        }
        System.out.print("Enter time signature (e.g., 4/4) [4/4]: ");
        String sigInput = scanner.nextLine().trim();
        if (sigInput.isEmpty()) sigInput = "4/4";
        String[] parts = sigInput.split("/");
        int beats = 4, unit = 4;
        if (parts.length == 2) {
            try {
                beats = Integer.parseInt(parts[0]);
                unit = Integer.parseInt(parts[1]);
                if (beats <= 0 || unit <= 0) { beats = 4; unit = 4; }
            } catch (NumberFormatException e) { beats = 4; unit = 4; }
        }
        scanner.close();
        Metronome metro = new Metronome(bpm, beats, unit);
        // Add shutdown hook to stop on Ctrl+C
        Runtime.getRuntime().addShutdownHook(new Thread(metro::stop));
        metro.start();
    }
}

File: metronome.go
// metronome.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

type Metronome struct {
	bpm              int
	beatsPerMeasure  int
	beatUnit         int
	running          bool
	interval         time.Duration
	beatCount        int
}

func NewMetronome(bpm, beats, unit int) *Metronome {
	return &Metronome{
		bpm:             bpm,
		beatsPerMeasure: beats,
		beatUnit:        unit,
		interval:        time.Duration(float64(60)/float64(bpm)*1000) * time.Millisecond,
	}
}

func (m *Metronome) beep() {
	switch runtime.GOOS {
	case "windows":
		// No simple beep without syscalls; we'll print
		fmt.Print(" 🔔")
	case "linux":
		cmd := exec.Command("beep", "-f", "800", "-l", "100")
		cmd.Run()
	default:
		fmt.Print(" 🔔")
	}
}

func (m *Metronome) visual(beatNum int, isDownbeat bool) {
	downbeatChar := "█"
	otherChar := "░"
	bar := downbeatChar
	if !isDownbeat {
		bar = otherChar
	}
	bar = strings.Repeat(bar, 10)
	fmt.Printf("\rBeat %d/%d: %s", beatNum, m.beatsPerMeasure, bar)
}

func (m *Metronome) tick() {
	m.beatCount++
	beatNum := (m.beatCount-1)%m.beatsPerMeasure + 1
	isDownbeat := beatNum == 1
	m.visual(beatNum, isDownbeat)
	m.beep()
}

func (m *Metronome) Start() {
	m.running = true
	m.beatCount = 0
	fmt.Printf("\nStarting metronome at %d BPM (%d/%d). Press Ctrl+C to stop.\n", m.bpm, m.beatsPerMeasure, m.beatUnit)
	ticker := time.NewTicker(m.interval)
	defer ticker.Stop()
	go func() {
		for m.running {
			<-ticker.C
			m.tick()
		}
	}()
	// Wait for interrupt
	ch := make(chan os.Signal, 1)
	signal.Notify(ch, os.Interrupt)
	<-ch
	m.running = false
	fmt.Println("\nStopped.")
}

func parseTimeSignature(sig string) (int, int) {
	parts := strings.Split(sig, "/")
	if len(parts) == 2 {
		b, err1 := strconv.Atoi(parts[0])
		u, err2 := strconv.Atoi(parts[1])
		if err1 == nil && err2 == nil && b > 0 && u > 0 {
			return b, u
		}
	}
	return 4, 4
}

func main() {
	fmt.Println("=== Metronome ===")
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter BPM (20-300) [120]: ")
	bpmStr, _ := reader.ReadString('\n')
	bpmStr = strings.TrimSpace(bpmStr)
	bpm := 120
	if bpmStr != "" {
		if v, err := strconv.Atoi(bpmStr); err == nil && v >= 20 && v <= 300 {
			bpm = v
		}
	}
	fmt.Print("Enter time signature (e.g., 4/4) [4/4]: ")
	sigStr, _ := reader.ReadString('\n')
	sigStr = strings.TrimSpace(sigStr)
	if sigStr == "" {
		sigStr = "4/4"
	}
	beats, unit := parseTimeSignature(sigStr)
	metro := NewMetronome(bpm, beats, unit)
	metro.Start()
}
